src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonDB.java

```diff
@@ -0,0 +1,361 @@
+            else if (left.isBefore(right))
+                return -1;
+            else if (left.isPast(right))
+                return 1;
+            else
+                return 0;
+        };
+    }
+
+    @Override
+    public GenomeLoc exon(final GenomeLoc overlapRegio) {
+        final int searchIndex = index(overlapRegio);
+        return searchIndex < 0 ? null : sortedIntervals.get(searchIndex);
+    }
+
+    @Override
+    public int index(final GenomeLoc location) {
+        if (location == GenomeLoc.WHOLE_GENOME)
+            return indexWholeGenome(location);
+        final int searchIndex = cachedBinarySearch(location);
+        if (searchIndex < 0)
+            return searchIndex;
+        else if (location.size() <= 1)
+            return searchIndex;
+        else if (searchIndex > 0 && !sortedIntervals.get(searchIndex - 1).isBefore(location))
+            throw new AmbiguousExonException("" + location + " overlaps with at least 2: "
+                    + sortedIntervals.get(searchIndex) + " and " + sortedIntervals.get(searchIndex - 1));
+        else if (searchIndex < sortedIntervals.size() - 1 &&
+                !sortedIntervals.get(searchIndex + 1).isPast(location))
+            throw new AmbiguousExonException("" + location + " overlaps with at least 2: "
+                    + sortedIntervals.get(searchIndex) + " and " + sortedIntervals.get(searchIndex + 1));
+        else
+            return searchIndex;
+    }
+
+    /**
+     * Special case for {@link #index(GenomeLoc)} for {@link GenomeLoc#WHOLE_GENOME}
+     * @param location the whole-genome location.
+     * @return never {@code null}.
+     */
+    private int indexWholeGenome(final GenomeLoc location) {
+        final int size = sortedIntervals.size();
+        if (size == 0)
+            return -1;
+        else if (size == 1)
+            return 0;
+        else
+            throw new AmbiguousExonException("" + location + " overlaps with at least 2: "
+                    + sortedIntervals.get(0) + " and " + sortedIntervals.get(1));
+    }
+
+    @Override
+    public List<GenomeLoc> exons(final GenomeLoc overlapRegion) {
+        if (overlapRegion == GenomeLoc.WHOLE_GENOME)
+            return Collections.unmodifiableList(sortedIntervals);
+        final int searchIndex = cachedBinarySearch(overlapRegion);
+        if (searchIndex < 0)
+            return Collections.emptyList();
+        final int firstMatchingIndex = extendSearchIndexBackwards(overlapRegion, searchIndex);
+        final int lastMatchingIndex = extendSearchIndexForward(overlapRegion, searchIndex);
+        if (firstMatchingIndex == lastMatchingIndex)
+            return Collections.singletonList(sortedIntervals.get(searchIndex));
+        else
+            return Collections.unmodifiableList(sortedIntervals.subList(firstMatchingIndex, lastMatchingIndex + 1));
+    }
+
+    private int lastBinarySearchResult = -1;
+
+    private int cachedBinarySearch(final GenomeLoc location) {
```

Comment:
can you add a comment about it? every reader of this code will ask the same question i did so a comment will help them understand why the code looks this way.